src/sources/kafka.rs

```diff
@@ -374,105 +390,406 @@ impl SourceConfig for KafkaSourceConfig {
+                    if let Ok(tpl) = consumer.assignment() {
+                        tpl.elements()
+                            .iter()
+                            .for_each(|el| {
+
+                            let tp: TopicPartition = (el.topic().into(), el.partition());
+                            if let Some(end) = end_signals.remove(&tp) {
+                                let _ = end.send(());
+                            }
+                            partition_state.revoke_partition(tp);
+                        });
+                    }
+
+                    drain_deadline = tokio::spawn(revoke_timeout(max_drain_ms));
+                    if partition_state.is_drain_complete() {
+                        partition_state.clear();
+                        drop(drain);
+                    } else if drain_signal.replace(drain).is_some() {
+                        debug!("Kafka consumer shutting down mid-rebalance.");
+                    }
 
-                    parse_message(msg, decoder.clone(), config.keys(), &finalizer, &mut out, &consumer, log_namespace).await;
+                    callbacks.close();
+                },
+            },
+
+            _ = &mut drain_deadline, if drain_signal.is_some() => {
+                debug!("Acknowledgement drain deadline reached. Dropping any pending ack streams for revoked partitions.");
+                for tp in partition_state.expect_drain.drain() {
+                    if let Some(handle) = abort_handles.remove(&tp) {
+                        handle.abort();
+                    }
+                }
+                partition_state.clear();
+
+                if let Err(e) = drain_signal.take().unwrap().send(()) {
+                    warn!("Error sending to drain signal: {}.", e);
                 }
             },
+
+            // Consumers are done, and callback channel is closed
+            else => {
+                break
+            }
         }
     }
+}
 
-    // Since commits are async internally, we try one last sync commit inside the interval
-    // in case there have been acks.
-    if let Ok(current_assignment) = consumer.assignment() {
-        // not logging on error because it will error if there are no offsets stored for a partition,
-        // and this is best-effort cleanup anyway
-        _ = consumer.commit(&current_assignment, CommitMode::Sync);
-    }
-    Ok(())
+fn drive_kafka_consumer(
+    consumer: Arc<StreamConsumer<KafkaSourceContext>>,
+    mut shutdown: ShutdownSignal,
+    mut eof: tokio::sync::oneshot::Receiver<()>,
+) {
+    Handle::current().block_on(async move {
+        let mut stream = consumer.stream();
+        loop {
+            tokio::select! {
+                _ = &mut shutdown => {
+                    consumer.context().shutdown();
+                    break
+                },
+
+                _ = &mut eof => {
+                    consumer.context().shutdown();
+                    break
+                },
```

Comment:
I believe we will only want to select on `_ = &mut eof` if the feature is enabled. Otherwise, the channel sender is dropped after the source is built, which would then trigger a shutdown, right? We can accomplish this with `futures::future::OptionFuture`.

```suggestion
fn drive_kafka_consumer(
    consumer: Arc<StreamConsumer<KafkaSourceContext>>,
    mut shutdown: ShutdownSignal,
    eof: Option<tokio::sync::oneshot::Receiver<()>>,
) {
    Handle::current().block_on(async move {
        let mut eof: OptionFuture<_> = eof.into();
        let mut stream = consumer.stream();
        loop {
            tokio::select! {
                _ = &mut shutdown => {
                    consumer.context().shutdown();
                    break
                },

                Some(_) = &mut eof => {
                    consumer.context().shutdown();
                    break
                },
```